<!DOCTYPE html>
<html lang="en">
<head>
    <meta charset="UTF-8">
    <meta http-equiv="X-UA-Compatible" content="IE=edge">
    <meta name="viewport" content="width=device-width, initial-scale=1.0">
    <title>Asyns-Await</title>
</head>
<body>
    <h1>Asyns-Await</h1>
    <script>

         function cuadradoPromise(value){
            if(typeof value !=="number"){
                return Promise.reject(`Error, el valor ${value} ingresado no es un numero`);
            }
            return new Promise((resolve,reject)=>{
                setTimeout(() => {
                    resolve({
                        value,
                        result:value*value
                    });
                }, 0 | Math.random()*2000);
            });
        }

        //Declarar funcion Asincrona Declarada
        async function functionAsincronaDeclarada (){
            //Manejo de errores
            try {
                console.log("Inicio de Async Function");
                //Porque la funcion del cuadrado devuelve un objeto, y esta variable va a ejecutar la funcion 
                //Inicio de la estructura Async
                let obj = await cuadradoPromise(0); 
                console.log(`Async Functions:${obj.value},${obj.result}`);
                
                obj = await cuadradoPromise(1); 
                console.log(`Async Functions:${obj.value},${obj.result}`);

                obj = await cuadradoPromise(2); 
                console.log(`Async Functions:${obj.value},${obj.result}`);

                obj = await cuadradoPromise(3); 
                console.log(`Async Functions:${obj.value},${obj.result}`);

                obj = await cuadradoPromise(4); 
                console.log(`Async Functions:${obj.value},${obj.result}`);

                obj = await cuadradoPromise(5); 
                console.log(`Async Functions:${obj.value},${obj.result}`);

                obj = await cuadradoPromise(6); 
                console.log(`Async Functions:${obj.value},${obj.result}`);

                console.log("Fin Async Function");


            } catch (err) {
                console.error(err);
            }
        }

        functionAsincronaDeclarada();


        //Funcion Expresada Async

        const functionAsincronaExpresada = async() =>{
             //Manejo de errores
             try {
                console.log("Inicio de Async Function");
                //Porque la funcion del cuadrado devuelve un objeto, y esta variable va a ejecutar la funcion 
                //Inicio de la estructura Async
                let obj = await cuadradoPromise(6); 
                console.log(`Async Functions:${obj.value},${obj.result}`);
                
                obj = await cuadradoPromise(7); 
                console.log(`Async Functions:${obj.value},${obj.result}`);

                obj = await cuadradoPromise(8); 
                console.log(`Async Functions:${obj.value},${obj.result}`);

                obj = await cuadradoPromise(9); 
                console.log(`Async Functions:${obj.value},${obj.result}`);

                obj = await cuadradoPromise(10); 
                console.log(`Async Functions:${obj.value},${obj.result}`);

                obj = await cuadradoPromise(11); 
                console.log(`Async Functions:${obj.value},${obj.result}`);

                obj = await cuadradoPromise(12); 
                console.log(`Async Functions:${obj.value},${obj.result}`);

                console.log("Fin Async Function");


            } catch (err) {
                console.error(err);
            }
        }
        functionAsincronaExpresada();
    </script>
</body>
</html>